README.md
# dart-buffer

This package generated for Buffer Utils. Package have reader and writer classes.

## Installation
        dart pub add dart_buffer

## Example
```dart
import 'dart:typed_data';

import 'package:dart_buffer/dart_buffer.dart';

void main() {
  final buffer = ByteData(8);
  final writer = BufferWriter(buffer);
  final reader = BufferReader(buffer);

  writer.setInt32(50332161);

  print(reader.getUInt32());
}
```

## LICENSE [MIT](LICENSE)


CHANGELOG.md
## 2.0.2

- update year in LICENSE 

## 2.0.1

- update LICENSE

## 2.0.0

- update license

## 2.0.1

- add new methods
- update convert methods

## 1.0.3

- fixed array bug

## 1.0.2

- offset increment bug fixed
- update tests
- add encodingLength function

## 1.0.1

- add getUInt, setUInt method

## 1.0.0

- Initial version.


test/common_helper.dart
int encodingLength(int number) {
  if (number < 0) {
    throw ArgumentError.value(
        number, 'number', 'Value should be unsigned integer.');
  }

  return (number < 0xfd
      ? 1
      : number <= 0xffff
          ? 3
          : number <= 0xffffffff
              ? 5
              : 9);
}


test/buffer_reader_test.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:dart_buffer/dart_buffer.dart';
import 'package:test/test.dart';

import 'common_helper.dart';

void testValue(BufferReader bufferReader, dynamic value, dynamic expectedValue,
    [int? expectedOffset]) {
  expectedOffset ??= expectedValue is List<int> ? expectedValue.length : 0;

  expect(bufferReader.offset, expectedOffset);

  if (expectedValue is List<int> && value is List<int>) {
    expect(
      value.sublist(
          0, value.length >= expectedOffset ? expectedOffset : value.length),
      expectedValue.sublist(
          0,
          expectedValue.length >= expectedOffset
              ? expectedOffset
              : expectedValue.length),
    );
  } else {
    expect(value as int, expectedValue);
  }
}

void main() {
  test('BufferReader.getInt8()', () {
    final values = [-128, -127, -1, 0, 1, 0x7d, 0x7f];
    final buffer = Int8List.fromList(values).buffer.asByteData();
    final bufferReader = BufferReader(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 1;
      final val = bufferReader.getInt8();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getUInt8()', () {
    final values = [0, 1, 0xfe, 0xff];
    final buffer = Uint8List.fromList(values).buffer.asByteData();
    final bufferReader = BufferReader(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 1;
      final val = bufferReader.getUInt8();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getInt16()', () {
    final values = [-32768, -32767, -1, 0, 1, 0x7ffd, 0x7fff];
    final buffer = Int16List.fromList(values).buffer.asByteData();
    final bufferReader = BufferReader(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 2;
      final val = bufferReader.getInt16();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getUInt16()', () {
    final values = [0, 1, 0x7ffd, 0x7fff];
    final buffer = Uint16List.fromList(values).buffer.asByteData();
    final bufferReader = BufferReader(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 2;
      final val = bufferReader.getUInt16();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getInt32()', () {
    final maxInt32 = 2147483647;
    final minInt32 = -2147483648;
    final values = <int>[
      minInt32,
      minInt32 ~/ 2,
      -1,
      0,
      1,
      maxInt32 ~/ 2,
      maxInt32
    ];
    final buffer = Int32List.fromList(values);
    final bufferReader = BufferReader.fromTypedData(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 4;
      final val = bufferReader.getInt32();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getUInt32()', () {
    final maxInt32 = 2147483647;
    final values = <int>[0, 1, maxInt32 ~/ 2, maxInt32];
    final buffer = Uint32List.fromList(values);
    final bufferReader = BufferReader.fromTypedData(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 4;
      final val = bufferReader.getUInt32();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getInt64()', () {
    final maxInt64 = 9223372036854775807;
    final minInt64 = -9223372036854775808;
    final values = <int>[
      minInt64,
      minInt64 ~/ 2,
      -1,
      0,
      1,
      maxInt64 ~/ 2,
      maxInt64
    ];
    final buffer = Int64List.fromList(values);
    final bufferReader = BufferReader.fromTypedData(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 8;
      final val = bufferReader.getInt64();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getUInt64()', () {
    final maxInt64 = 9223372036854775807;
    final values = <int>[0, 1, maxInt64 ~/ 2, maxInt64];
    final buffer = Uint64List.fromList(values);
    final bufferReader = BufferReader.fromTypedData(buffer);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + 8;
      final val = bufferReader.getUInt64();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getVarInt()', () {
    final hexStr =
        '0001fcfdfd00fdfe00fdff00fd0001fdfefffdfffffe00000100fefefffffffeffffffffff0000000001000000ffffffffffffff1f00';
    final values = [
      0,
      1,
      252,
      253,
      254,
      255,
      256,
      pow(2, 16).toInt() - 2,
      pow(2, 16).toInt() - 1,
      pow(2, 16).toInt(),
      pow(2, 32).toInt() - 2,
      pow(2, 32).toInt() - 1,
      pow(2, 32).toInt(),
      9007199254740991,
    ];
    final bufferReader = BufferReader.fromHex(hexStr);
    for (var v in values) {
      int expectedOffset = bufferReader.offset;
      final val = bufferReader.getVarInt();
      expectedOffset += encodingLength(val);
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getSlice()', () {
    final hexStr = '0101020304feff';
    final values = [
      [1],
      [1, 2, 3, 4],
      [254, 255]
    ];
    final bufferReader = BufferReader.fromHex(hexStr);
    for (var v in values) {
      final expectedOffset = bufferReader.offset + v.length;
      final val = bufferReader.getSlice(v.length).buffer.asUint8List();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
  test('BufferReader.getVarSlice()', () {
    const hexStr =
        '0101fc020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202fdfd0003030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303';

    final values = [
      Uint8List.fromList(List.filled(1, 1)),
      Uint8List.fromList(List.filled(252, 2)),
      Uint8List.fromList(List.filled(253, 3))
    ];

    final bufferReader = BufferReader.fromHex(hexStr);
    for (var v in values) {
      final expectedOffset =
          bufferReader.offset + encodingLength(v.length) + v.length;
      final val = bufferReader.getVarSlice().buffer.asUint8List();
      testValue(bufferReader, val, v, expectedOffset);
    }
  });
}


test/buffer_writer_test.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_buffer/dart_buffer.dart';
import 'package:test/test.dart';

import 'common_helper.dart';

void main() {
  void testBuffer<T extends List<int>>(
      BufferWriter bufferWriter, T expectedBuffer,
      [int? expectedOffset]) {
    expectedOffset ??= expectedBuffer.length;
    expect(bufferWriter.offset, equals(expectedOffset));
    expect(
      bufferWriter.toTypeData<T>().sublist(0, expectedOffset),
      expectedBuffer.sublist(0, expectedOffset),
    );
  }

  test('BufferWriter.withCapacity()', () {
    Uint8List expectedBuffer = (hex.decode('04030201') as Uint8List);
    BufferWriter withCapacity = BufferWriter.withCapacity(4);
    withCapacity.setInt32(0x04030201, Endian.big);

    testBuffer(withCapacity, expectedBuffer);
  });
  test('BufferWriter.setInt8()', () {
    final values = [-128, -127, -1, 0, 1, 0x7e, 0x7d];
    final expectedBuffer = Int8List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in values) {
      final expectedOffset = bufferWriter.offset + 1;
      bufferWriter.setInt8(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(bufferWriter.buffer.getInt8(bufferWriter.offset - 1), equals(v));
    }
  });
  test('BufferWriter.setUInt8()', () {
    final values = [0, 1, 0xfe, 0xff];
    final expectedBuffer = Uint8List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 1;
      bufferWriter.setUInt8(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(bufferWriter.buffer.getUint8(bufferWriter.offset - 1), equals(v));
    }
  });
  test('BufferWriter.setInt16()', () {
    final values = [-32768, -32767, -1, 0, 1, 0x7ffd, 0x7fff];
    final expectedBuffer = Int16List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 2;
      bufferWriter.setInt16(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getInt16(bufferWriter.offset - 2, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setUInt16()', () {
    final values = [0, 1, 0x7ffd, 0x7fff];
    final expectedBuffer = Uint16List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 2;
      bufferWriter.setUInt16(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getUint16(bufferWriter.offset - 2, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setInt32()', () {
    final maxInt32 = 2147483647;
    final minInt32 = -2147483648;
    final values = <int>[
      minInt32,
      minInt32 ~/ 2,
      -1,
      0,
      1,
      maxInt32 ~/ 2,
      maxInt32
    ];
    final expectedBuffer = Int32List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 4;
      bufferWriter.setInt32(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getInt32(bufferWriter.offset - 4, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setUInt32()', () {
    final maxInt32 = 2147483647;
    final values = <int>[0, 1, maxInt32 ~/ 2, maxInt32];
    final expectedBuffer = Uint32List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 4;
      bufferWriter.setUInt32(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getUint32(bufferWriter.offset - 4, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setInt64()', () {
    final maxInt64 = 9223372036854775807;
    final minInt64 = -9223372036854775808;
    final values = <int>[
      minInt64,
      minInt64 ~/ 2,
      -1,
      0,
      1,
      maxInt64 ~/ 2,
      maxInt64
    ];
    final expectedBuffer = Int64List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 8;
      bufferWriter.setInt64(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getInt64(bufferWriter.offset - 8, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setUInt64()', () {
    final maxInt64 = 9223372036854775807;
    final values = <int>[0, 1, maxInt64 ~/ 2, maxInt64];
    final expectedBuffer = Uint64List.fromList(values);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var v in expectedBuffer) {
      final expectedOffset = bufferWriter.offset + 8;
      bufferWriter.setUInt64(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(
          bufferWriter.buffer.getUint64(bufferWriter.offset - 8, Endian.little),
          equals(v));
    }
  });
  test('BufferWriter.setVarInt()', () {
    final expectedBuffer = hex.decode(
        '0001fcfdfd00fdfe00fdff00fd0001fdfefffdfffffe00000100fefefffffffeffffffffff0000000001000000ffffffffffffff1f00');
    final values = [
      0,
      1,
      252,
      253,
      254,
      255,
      256,
      pow(2, 16).toInt() - 2,
      pow(2, 16).toInt() - 1,
      pow(2, 16).toInt(),
      pow(2, 32).toInt() - 2,
      pow(2, 32).toInt() - 1,
      pow(2, 32).toInt(),
      9007199254740991,
    ];
    final byteData = ByteData(expectedBuffer.length);
    final bufferWriter = BufferWriter(byteData);
    final bufferReader = BufferReader(byteData);
    for (var v in values) {
      int expectedOffset = bufferWriter.offset + encodingLength(v);
      bufferWriter.setVarInt(v);
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(bufferReader.getVarInt(), equals(v));
    }
  });
  test('BufferWriter.setSlice()', () {
    final List<List<int>> values = [
      [],
      [1],
      [1, 2, 3, 4],
      [254, 255]
    ];
    final expectedBuffer = Uint8List.fromList([1, 1, 2, 3, 4, 254, 255]);
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.length));
    for (var v in values) {
      final expectedOffset = bufferWriter.offset + v.length;
      bufferWriter.setSlice(Uint8List.fromList(v));
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(bufferWriter.toTypeData<Uint8List>().sublist(0, expectedOffset),
          expectedBuffer.sublist(0, expectedOffset));
    }
    expect(bufferWriter.offset, equals(expectedBuffer.lengthInBytes));
    expect(
        bufferWriter.toTypeData<Uint8List>().sublist(0, expectedBuffer.length),
        expectedBuffer.sublist(0, expectedBuffer.lengthInBytes));
    try {
      bufferWriter.setSlice(Uint8List.fromList([0, 0]));
    } catch (e) {
      expect(e.toString(),
          matches(RegExp(r'^Exception: Cannot set slice out of bounds$')));
      return;
    }
    throw Exception('Should be throwable.');
  });
  test('BufferWriter.setVarSlice()', () {
    final values = [
      Uint8List.fromList(List.filled(1, 1)),
      Uint8List.fromList(List.filled(252, 2)),
      Uint8List.fromList(List.filled(253, 3))
    ];
    final expectedBuffer = Uint8List.fromList(Uint8List.fromList([0x01, 0x01]) +
        Uint8List.fromList([0xfc]) +
        Uint8List.fromList(List.filled(252, 0x02)) +
        Uint8List.fromList([0xfd, 0xfd, 0x00]) +
        Uint8List.fromList(List.filled(253, 0x03)));
    final bufferWriter = BufferWriter(ByteData(expectedBuffer.lengthInBytes));
    for (var value in values) {
      final expectedOffset =
          bufferWriter.offset + encodingLength(value.length) + value.length;
      bufferWriter.setVarSlice(Uint8List.fromList(value));
      expect(bufferWriter.offset, equals(expectedOffset));
      expect(bufferWriter.toTypeData<Uint8List>().sublist(0, expectedOffset),
          expectedBuffer.sublist(0, expectedOffset));
    }
    expect(bufferWriter.offset, equals(expectedBuffer.lengthInBytes));
    expect(
        bufferWriter.toTypeData<Uint8List>().sublist(0, expectedBuffer.length),
        expectedBuffer.sublist(0, expectedBuffer.lengthInBytes));
  });
  test('BufferWriter.end()', () {
    Uint8List expectedBuffer = (hex.decode('0403020108070605') as Uint8List);
    BufferWriter bufferWriter = BufferWriter.withCapacity(8);
    bufferWriter.setUInt32(0x01020304, Endian.little);
    bufferWriter.setUInt32(0x05060708, Endian.little);

    final result = bufferWriter.end();
    testBuffer(bufferWriter, result.buffer.asUint8List());

    testBuffer(bufferWriter, expectedBuffer);
  });
}


lib/src/buffer_reader.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';

/// Buffer Reader
class BufferReader {
  ByteData buffer;
  int offset;

  get length => buffer.lengthInBytes;

  /// BufferReader from ByteData
  BufferReader(this.buffer, [this.offset = 0]);

  /// BufferReader from TypedData
  factory BufferReader.fromTypedData(TypedData buffer, [offset = 0]) {
    return BufferReader(buffer.buffer.asByteData(), offset);
  }

  /// BufferReader from hex string
  factory BufferReader.fromHex(String data, [offset = 0]) {
    return BufferReader.fromTypedData(
        Uint8List.fromList(hex.decode(data)), offset);
  }

  /// Set offset to zero
  reset() {
    offset = 0;
  }

  ///Reads byteLength number of bytes from buf at the current offset
  /// and interprets the result as an unsigned, [Endian] integer
  /// supporting up to 48 bits of accuracy.
  int getUInt([int byteLength = 0, Endian endian = Endian.little]) {
    if (offset + byteLength > length) {
      throw IndexError(byteLength, buffer, 'IndexError',
          'Cannot get UIntLE out of bounds', buffer.lengthInBytes);
    }
    if (endian != Endian.little) {
      var val = buffer.buffer.asUint8List()[offset];
      var mul = 1;
      var i = 0;
      while (++i < byteLength && (mul *= 0x100) != 0) {
        val += buffer.buffer.asUint8List()[offset + i] * mul;
      }
      return val;
    } else {
      var val = buffer.buffer.asUint8List()[offset + --byteLength];
      var mul = 1;
      while (byteLength > 0 && (mul *= 0x100) != 0) {
        val += buffer.buffer.asUint8List()[offset + --byteLength] * mul;
      }
      return val;
    }
  }

  /// Returns the (possibly negative) integer represented by the byte at
  /// current offset in buffer, in two's complement binary
  /// representation.
  ///
  /// The return value will be between -128 and 127, inclusive.
  int getInt8() {
    final result = buffer.getInt8(offset);
    offset++;
    return result;
  }

  /// Returns the positive integer represented by the byte.
  /// Increments the offset value by one.
  /// The return value will be between 0 and 255, inclusive.
  int getUInt8() {
    final result = buffer.getUint8(offset);
    offset++;
    return result;
  }

  /// Returns the (possibly negative) integer represented by the four bytes
  /// Increments the offset value by four
  /// The return value will be between -2<sup>15</sup> and 2<sup>15</sup> - 1,
  /// inclusive.
  int getInt16([Endian endian = Endian.little]) {
    final result = buffer.getInt16(offset, endian);
    offset += 2;
    return result;
  }

  /// Returns the positive integer represented by the four bytes
  /// Increments the offset value by four
  /// The return value will be between 0 and  2<sup>16</sup> - 1, inclusive.
  int getUInt16([Endian endian = Endian.little]) {
    final result = buffer.getUint16(offset, endian);
    offset += 2;
    return result;
  }

  /// Returns the (possibly negative) integer represented by the four bytes
  /// Increments the offset value by four
  /// The return value will be between -2<sup>15</sup> and 2<sup>15</sup> - 1,
  /// inclusive.
  int getInt32([Endian endian = Endian.little]) {
    final result = buffer.getInt32(offset, endian);
    offset += 4;
    return result;
  }

  /// Returns the positive integer represented by the four bytes
  /// Increments the offset value by four
  /// The return value will be between 0 and  2<sup>32</sup> - 1, inclusive.
  int getUInt32([Endian endian = Endian.little]) {
    final result = buffer.getUint32(offset, endian);
    offset += 4;
    return result;
  }

  /// Returns the positive integer represented by the eight bytes
  /// Increments the offset value by eight
  /// The return value will be between 0 and  2<sup>64</sup> - 1, inclusive.
  int getInt64([Endian endian = Endian.little]) {
    final result = buffer.getInt64(offset, endian);
    offset += 8;
    return result;
  }

  /// Returns the positive integer represented by the eight bytes
  /// Increments the offset value by eight
  /// The return value will be between 0 and  2<sup>64</sup> - 1, inclusive.
  int getUInt64([Endian endian = Endian.little]) {
    final result = buffer.getUint64(offset, endian);
    offset += 8;
    return result;
  }

  /// Returns the floating point number represented by the four bytes,
  /// in IEEE 754 single-precision binary floating-point format (binary32).
  double getFloat32([Endian endian = Endian.little]) {
    final result = buffer.getFloat32(offset, endian);
    offset += 4;
    return result;
  }

  /// Returns the floating point number represented by the eight bytes,
  /// in IEEE 754 double-precision binary floating-point format (binary64).
  double getFloat64([Endian endian = Endian.little]) {
    final result = buffer.getFloat64(offset, endian);
    offset += 8;
    return result;
  }

  /// Get next byte and return an int value based on the data length specified by this byte.
  /// The return value will between 0 and 2<sup>64</sup> - 1, inclusive.
  int getVarInt([Endian endian = Endian.little]) {
    ByteData bufferByteData = buffer.buffer.asByteData();
    int first = bufferByteData.getUint8(offset);
    int output;
    // 8 bit
    if (first < 0xfd) {
      output = first;
      offset += 1;
      // 16 bit
    } else if (first == 0xfd) {
      output = bufferByteData.getUint16(offset + 1, endian);
      offset += 3;

      // 32 bit
    } else if (first == 0xfe) {
      output = bufferByteData.getUint32(offset + 1, endian);
      offset += 5;

      // 64 bit
    } else {
      output = bufferByteData.getUint64(offset + 1, endian);
      offset += 9;
    }

    return output;
  }

  /// Get next n bytes
  /// The return value will be ByteData(n)
  ByteData getSlice(int n) {
    if (buffer.lengthInBytes < offset + n) {
      throw IndexError(n, buffer, 'IndexError',
          'Cannot get slice out of bounds', buffer.lengthInBytes);
    }
    final result = buffer.buffer
        .asUint8List()
        .sublist(offset, offset + n)
        .buffer
        .asByteData();
    offset += n;
    return result;
  }

  /// Get next byte and return an ByteData value based on the data length specified by this byte.
  /// The return value will be ByteData.
  ByteData getVarSlice([Endian endian = Endian.little]) {
    return getSlice(getVarInt(endian));
  }

  /// Get vector list
  /// The return value will be List<ByteData>.
  List<ByteData> getVector([Endian endian = Endian.little]) {
    final count = getVarInt();
    List<ByteData> vector = [];
    for (int i = 0; i < count; i++) {
      vector.add(getVarSlice(endian));
    }
    return vector;
  }
}


lib/src/buffer_writer.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:convert/convert.dart';

/// Buffer Writer
class BufferWriter {
  ByteData buffer;
  int offset;

  get length => buffer.lengthInBytes;

  /// BufferWriter from ByteData
  BufferWriter(this.buffer, [this.offset = 0]);

  /// BufferWriter from TypedData
  factory BufferWriter.fromTypedData(TypedData buffer, [offset = 0]) {
    return BufferWriter(buffer.buffer.asByteData(), offset);
  }

  /// BufferWriter from hex string
  factory BufferWriter.fromHex(String data, [offset = 0]) {
    return BufferWriter.fromTypedData(
        Uint8List.fromList(hex.decode(data)), offset);
  }

  /// BufferWriter with capacity
  factory BufferWriter.withCapacity(int size) {
    return BufferWriter(ByteData(size));
  }

  /// Set offset to zero
  reset() {
    offset = 0;
  }

  ///Writes byteLength bytes of value to buf at the current
  /// offset as [Endian]. Supports up to 48 bits of accuracy.
  void setUInt(int value, [int byteLength = 0, Endian endian = Endian.little]) {
    num maxBytes = math.pow(2, 8 * byteLength) - 1;

    if (value > maxBytes || value < 0) {
      throw RangeError('"value" argument is out of bounds');
    }
    if (offset + byteLength > buffer.lengthInBytes) {
      throw RangeError('"byteLength" out of range');
    }
    int mul = 1;
    if (endian == Endian.little) {
      int i = 0;
      buffer.buffer.asUint8List()[offset] = value & 0xFF;
      while (++i < byteLength && (mul *= 0x100) != 0) {
        buffer.buffer.asUint8List()[offset + i] = value ~/ mul & 0xFF;
      }
    } else {
      int i = byteLength - 1;
      buffer.buffer.asUint8List()[offset + i] = value & 0xFF;
      while (--i >= 0 && (mul *= 0x100) != 0) {
        buffer.buffer.asUint8List()[offset + i] = value ~/ mul & 0xFF;
      }
    }
    offset = offset + byteLength;
  }

  /// Sets the byte at the current offset in buffer to the
  /// two's complement binary representation of the specified [value], which
  /// must fit in a single byte.
  ///
  /// In other words, [value] must be between -128 and 127, inclusive.
  void setInt8(int i) {
    buffer.setInt8(offset, i);
    offset++;
  }

  /// Sets the byte at the current offset in buffer to the
  /// unsigned binary representation of the specified [value], which must fit
  /// in a single byte.
  ///
  /// In other words, [value] must be between 0 and 255, inclusive.
  void setUInt8(int i) {
    buffer.setUint8(offset, i);
    offset++;
  }

  /// Sets the two bytes starting at the current offset in buffer
  /// to the two's complement binary representation of the specified
  /// [value], which must fit in two bytes.
  ///
  /// In other words, [value] must lie
  /// between -2<sup>15</sup> and 2<sup>15</sup> - 1, inclusive.
  void setInt16(int i, [Endian endian = Endian.little]) {
    buffer.setInt16(offset, i, endian);
    offset += 2;
  }

  /// Sets the two bytes starting at the current offset in this object
  /// to the unsigned binary representation of the specified [value],
  /// which must fit in two bytes.
  ///
  /// In other words, [value] must be between
  /// 0 and 2<sup>16</sup> - 1, inclusive.
  void setUInt16(int i, [Endian endian = Endian.little]) {
    buffer.setUint16(offset, i, endian);
    offset += 2;
  }

  /// Sets the four bytes starting at the current offset in buffer
  /// to the two's complement binary representation of the specified
  /// [value], which must fit in four bytes.
  ///
  /// In other words, [value] must lie
  /// between -2<sup>31</sup> and 2<sup>31</sup> - 1, inclusive.
  void setInt32(int i, [Endian endian = Endian.little]) {
    buffer.setInt32(offset, i, endian);
    offset += 4;
  }

  /// Sets the four bytes starting at the current offset in buffer
  /// to the unsigned binary representation of the specified [value],
  /// which must fit in four bytes.
  ///
  /// In other words, [value] must be between
  /// 0 and 2<sup>32</sup> - 1, inclusive.
  void setUInt32(int i, [Endian endian = Endian.little]) {
    buffer.setUint32(offset, i, endian);
    offset += 4;
  }

  /// Sets the eight bytes starting at the current offset in buffer
  /// object to the two's complement binary representation of the specified
  /// [value], which must fit in eight bytes.
  ///
  /// In other words, [value] must lie
  /// between -2<sup>63</sup> and 2<sup>63</sup> - 1, inclusive.
  void setInt64(int i, [Endian endian = Endian.little]) {
    buffer.setInt64(offset, i, endian);
    offset += 8;
  }

  /// Sets the eight bytes starting at the current offset in buffer
  /// to the unsigned binary representation of the specified [value],
  /// which must fit in eight bytes.
  ///
  /// In other words, [value] must be between
  /// 0 and 2<sup>64</sup> - 1, inclusive.
  void setUInt64(int i, [Endian endian = Endian.little]) {
    buffer.setUint64(offset, i, endian);
    offset += 8;
  }

  /// Sets the four bytes starting at the current offset in buffer
  /// to the IEEE 754 single-precision binary floating-point
  /// (binary32) representation of the specified [value].
  ///
  /// **Note that this method can lose precision.** The input [value] is
  /// a 64-bit floating point value, which will be converted to 32-bit
  /// floating point value by IEEE 754 rounding rules before it is stored.
  /// If [value] cannot be represented exactly as a binary32, it will be
  /// converted to the nearest binary32 value.  If two binary32 values are
  /// equally close, the one whose least significant bit is zero will be used.
  /// Note that finite (but large) values can be converted to infinity, and
  /// small non-zero values can be converted to zero.
  void setFloat32(double i, [Endian endian = Endian.little]) {
    buffer.setFloat32(offset, i, endian);
    offset += 4;
  }

  /// Sets the eight bytes starting at the current offset in byte
  /// to the IEEE 754 double-precision binary floating-point
  /// (binary64) representation of the specified [value].
  void setFloat64(double i, [Endian endian = Endian.little]) {
    buffer.setFloat64(offset, i, endian);
    offset += 8;
  }

  /// Sets next byte [i] byte length and after then write [i]
  void setVarInt(int i, [Endian endian = Endian.little]) {
    if (i < 0) {
      throw RangeError('value out of range');
    }

    ByteData bufferByteData = buffer.buffer.asByteData();

    // 8 bit
    if (i < 0xfd) {
      bufferByteData.setUint8(offset, i);
      offset += 1;

      // 16 bit
    } else if (i <= 0xffff) {
      bufferByteData.setUint8(offset, 0xfd);
      bufferByteData.setUint16(offset + 1, i, endian);
      offset += 3;

      // 32 bit
    } else if (i <= 0xffffffff) {
      bufferByteData.setUint8(offset, 0xfe);
      bufferByteData.setUint32(offset + 1, i, endian);
      offset += 5;

      // 64 bit
    } else {
      bufferByteData.setUint8(offset, 0xff);
      bufferByteData.setUint64(offset + 1, i, endian);
      offset += 9;
    }
  }

  /// Sets next bytes
  void setSlice(TypedData slice) {
    if (buffer.lengthInBytes < offset + slice.lengthInBytes) {
      throw Exception('Cannot set slice out of bounds');
    }
    buffer.buffer.asUint8List().setAll(offset, slice.buffer.asUint8List());
    offset += slice.lengthInBytes;
  }

  /// Sets next byte and return an ByteData value based on the data length specified by this byte.
  void setVarSlice(TypedData slice) {
    setVarInt(slice.lengthInBytes);
    setSlice(slice);
  }

  /// Sets vector list
  void setVector(List<TypedData> vector) {
    setVarInt(vector.length);
    for (var buf in vector) {
      setVarSlice(buf);
    }
  }

  ByteData end() {
    if (buffer.lengthInBytes == offset) {
      return buffer;
    }
    throw Exception('buffer size ${buffer.lengthInBytes}, offset $offset');
  }

  /// Convert [ByteData] to [T]
  T toTypeData<T extends List<int>>() {
    switch (T) {
      case Uint8List:
        return buffer.buffer.asUint8List() as T;
      case Int8List:
        return buffer.buffer.asInt8List() as T;
      case Uint16List:
        return buffer.buffer.asUint16List() as T;
      case Int16List:
        return buffer.buffer.asInt16List() as T;
      case Uint32List:
        return buffer.buffer.asUint32List() as T;
      case Int32List:
        return buffer.buffer.asInt32List() as T;
      default:
        throw FormatException('Invalid Generic Type');
    }
  }
}
